exporter/clickhouseexporter/exporter.go

```diff
@@ -47,20 +51,161 @@ func newExporter(logger *zap.Logger, cfg *Config) (*clickhouseExporter, error) {
 }
 
 // Shutdown will shutdown the exporter.
-func (e *clickhouseExporter) Shutdown(ctx context.Context) error {
+func (e *clickhouseExporter) Shutdown(_ context.Context) error {
 	if e.client != nil {
 		return e.client.Close()
 	}
 	return nil
 }
 
 func (e *clickhouseExporter) pushLogsData(ctx context.Context, ld pdata.Logs) error {
-	// TODO implement write data to clickhouse
-	return nil
+	start := time.Now()
+	err := doWithTx(ctx, e.client, func(tx *sql.Tx) error {
+		statement, err := tx.PrepareContext(ctx, query)
+		if err != nil {
+			return fmt.Errorf("PrepareContext:%w", err)
+		}
+		defer func() {
+			_ = statement.Close()
+		}()
+		for i := 0; i < ld.ResourceLogs().Len(); i++ {
+			logs := ld.ResourceLogs().At(i)
+			res := logs.Resource()
+			resourceKeys, resourceValues := attributesToSlice(res.Attributes())
+			for j := 0; j < logs.InstrumentationLibraryLogs().Len(); j++ {
+				rs := logs.InstrumentationLibraryLogs().At(j).LogRecords()
+				for k := 0; k < rs.Len(); k++ {
+					r := rs.At(k)
+					attrKeys, attrValues := attributesToSlice(r.Attributes())
+					_, err = statement.ExecContext(ctx,
+						r.Timestamp().AsTime(),
+						r.TraceID().HexString(),
+						r.SpanID().HexString(),
+						r.Flags(),
+						r.SeverityText(),
+						r.SeverityNumber(),
+						r.Body().AsString(),
+						resourceKeys,
+						resourceValues,
+						attrKeys,
+						attrValues,
+					)
+					if err != nil {
+						return fmt.Errorf("ExecContext:%w", err)
+					}
+				}
+			}
+		}
+		return nil
+	})
+	duration := time.Since(start)
+	e.logger.Debug("insert logs", zap.Int("records", ld.LogRecordCount()),
+		zap.String("cost", duration.String()))
+	return err
+}
+
+func attributesToSlice(attributes pdata.AttributeMap) ([]string, []string) {
+	keys := make([]string, 0, attributes.Len())
+	values := make([]string, 0, attributes.Len())
+	attributes.Range(func(k string, v pdata.AttributeValue) bool {
+		keys = append(keys, formatKey(k))
+		values = append(values, v.AsString())
+		return true
+	})
+	return keys, values
+}
+
+func formatKey(k string) string {
+	return strings.ReplaceAll(k, ".", "_")
 }
 
+const (
+	// language=ClickHouse SQL
+	createLogsTableSQL = `
+CREATE TABLE IF NOT EXISTS logs_local (
+     Timestamp DateTime CODEC(Delta, ZSTD(1)),
+     TraceId String CODEC(ZSTD(1)),
+     SpanId String CODEC(ZSTD(1)),
+     TraceFlags UInt32,
+     SeverityText LowCardinality(String) CODEC(ZSTD(1)),
+     SeverityNumber Int32,
+     Body String CODEC(ZSTD(1)),
+     Attributes Nested
+     (
+         Key LowCardinality(String),
+         Value String
+     ) CODEC(ZSTD(1)),
+     Resource Nested
```

Comment:
I would recommend to call the column `ResourceAttributes`, but I leave it up to you

Reply:
This field name is align to the specification. I want to keep it. thanks. https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#log-and-event-record-definition

Reply:
Hm. The definition of the Resource is not correct in the doc. Resource its not the same as Log.Attributes. Resource has Attributes as a field: Resource.Attributes. Looks at the proto definition https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/resource/v1/resource.proto#L31

Reply:
It think it would be more clear to call them LogAttributes (or LogAttrs) and ResourceAttributes (or ResourceAttrs).

Reply:
I See. I can change
```
    Attributes Nested
        (
        Key LowCardinality(String),
        Value String
        ) CODEC(ZSTD(1)),
    Resource Nested
        (
        Key LowCardinality(String),
        Value String
        ) CODEC(ZSTD(1)),
```
to
```
    LogAttributes Nested
        (
        Key LowCardinality(String),
        Value String
        ) CODEC(ZSTD(1)),
    ResourceAttributes Nested
        (
        Key LowCardinality(String),
        Value String
        ) CODEC(ZSTD(1)),
```
, right?

Reply:
Yes, looks correct

Reply:
Done.